components/script/dom/selection.rs

```diff
@@ -0,0 +1,523 @@
+            if node.is_doctype() {
+                // Spec doesn't say to test this directly, but
+                // this would otherwise need to happen in steps 5-7
+                // as a rethrow from setting the range boundary
+                return Err(Error::InvalidNodeType);
+            }
+
+            if offset > node.len() {
+                // As with is_doctype, not explicit in selection spec steps
+                // implied by range-setting
+                return Err(Error::IndexSize);
+            }
+
+            // let is_old_anchor_before_or_equal;
+            // Step 4
+            if !self.is_same_root(&*range.StartContainer()) {
+                // Step 5, and its following 8 and 9
+                self.set_range(&*Range::new(&self.document, node, offset, node, offset));
+                self.direction.set(Direction::Forwards);
+            } else {
+                let old_anchor_node = &*self.GetAnchorNode().unwrap(); // has range, therefore has anchor node
+                let old_anchor_offset = self.AnchorOffset();
+                let is_old_anchor_before_or_equal = {
+                    if old_anchor_node == node {
+                        old_anchor_offset <= offset
+                    } else {
+                        old_anchor_node.is_before(node)
+                    }
+                };
+                if is_old_anchor_before_or_equal {
+                    // Step 6, and its following 8 and 9
+                    self.set_range(&*Range::new(
+                        &self.document,
+                        old_anchor_node,
+                        old_anchor_offset,
+                        node,
+                        offset,
+                    ));
+                    self.direction.set(Direction::Forwards);
+                } else {
+                    // Step 7, and its following 8 and 9
+                    self.set_range(&*Range::new(
+                        &self.document,
+                        node,
+                        offset,
+                        old_anchor_node,
+                        old_anchor_offset,
+                    ));
+                    self.direction.set(Direction::Backwards);
+                }
+            };
+        } else {
+            // Step 2
+            return Err(Error::InvalidState);
+        }
+        return Ok(());
+    }
+
+    // https://w3c.github.io/selection-api/#dom-selection-setbaseandextent
+    fn SetBaseAndExtent(
+        &self,
+        anchor_node: &Node,
+        anchor_offset: u32,
+        focus_node: &Node,
+        focus_offset: u32,
+    ) -> ErrorResult {
+        // Step 1
+        if anchor_offset > anchor_node.len() || focus_offset > focus_node.len() {
+            return Err(Error::IndexSize);
+        }
+
+        // Step 2
+        if !self.is_same_root(anchor_node) || !self.is_same_root(focus_node) {
+            return Ok(());
+        }
+
+        if anchor_node.is_doctype() || focus_node.is_doctype() {
```

Comment:
Looking at the other methods that are doing this, this should probably be before the IndexSize check, not after the same_root one.